prompts/internal/fstring/parser.go

```diff
@@ -0,0 +1,153 @@
+package fstring
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+type parser struct {
+	data   []rune
+	result []rune
+	idx    int
+	values map[string]any
+}
+
+func newParser(s string, values map[string]any) *parser {
+	if len(values) == 0 {
+		values = map[string]any{}
+	}
+	return &parser{
+		data:   []rune(s),
+		result: nil,
+		idx:    0,
+		values: values,
+	}
+}
+
+func (r *parser) parse() error {
+	for r.hasMore() {
+		existLeftCurlyBracket, tmp, err := r.scanToLeftCurlyBracket()
+		if err != nil {
+			return err
+		}
+		r.result = append(r.result, tmp...)
+		if !existLeftCurlyBracket {
+			continue
+		}
+
+		tmp, err = r.scanToRightCurlyBracket()
+		if err != nil {
+			return err
+		}
+		valName := strings.TrimSpace(string(tmp))
+		if valName == "" {
+			return fmt.Errorf("empty expression not allowed")
```

Comment:
please heed this lint warning and define an exported error value fo these conditions